File: python/src/tensor_op/cpp_reference/src/gen_kv_slc.cpp
/**

Copyright (c) 2025 Huawei Technologies Co., Ltd.
This file is a part of the CANN Open Software.
Licensed under CANN Open Software License Agreement Version 1.0 (the "License").
Please refer to the License for details. You may not use this file except in compliance with the License.
THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
See LICENSE in the root of the software repository for the full text of the License.
*/
/*!

\file page_attention.cpp
\brief
*/

#include "interface/function/function.h"
#include "interface/configs/config_manager.h"

using namespace npu::tile_fwk;

const int NUM1024 = 1024;

namespace npu::tile_fwk {
    

struct KvSlcTileShapeConfig {
    std::array<int, TILE_VEC_DIMS> v0TileShape;
}; 


void KvSlcCompute(Tensor &topK_indcies, Tensor &topK_tensor_shape, Tensor &kvNopeCache, Tensor &kRopeCache, Tensor &kvActSeqs, int front, int near, int topk, int l_prime,
                  int n2, Tensor &blockTable, int blockSize, Tensor &k_slcOut, Tensor &v_slcOut, Tensor &kvSlcActSeqs, KvSlcTileShapeConfig &tileConfig, bool debug = false) {
    auto v0Tile = tileConfig.v0TileShape;
    SymbolicScalar b = topK_indcies->shape[0];
    SymbolicScalar s = topK_indcies->shape[1];
    SymbolicScalar kv_lora_rank = kvNopeCache.GetShape(1) / n2;
    SymbolicScalar rope_dim = kRopeCache.GetShape(1) / n2;
    LOOP("LOOP_L0_batchIdx", FunctionType::DYNAMIC_LOOP, batchIdx, LoopRange(0, b, 1), {}, true) {
        SymbolicScalar curActSeq = GetTensorData(kvActSeqs, {batchIdx});
        LOOP("LOOP_L1_slcIdx", FunctionType::DYNAMIC_LOOP, slcIdx, LoopRange(0, s, 1)) {
            LOOP("LOOP_L2_kvSlcIdx", FunctionType::DYNAMIC_LOOP, nkvIdx, LoopRange(0, n2, 1)) {
                TileShape::Current().SetVecTile(v0Tile[0], v0Tile[1]);
                SymbolicScalar s_slc = GetTensorData(topK_tensor_shape, {batchIdx, slcIdx});
                SymbolicScalar positions = 0;
                SymbolicScalar prime_value = l_prime;
                SymbolicScalar slcSeqLen = 0;
                for (int topKIdx = 0; topKIdx < topk; topKIdx++) {
                    // 获取到topk的position
                    // 头部的front个
                    if (topKIdx < front) {
                        positions = topKIdx * l_prime;
                    } else if (topKIdx > (topk - near - front)) {
                    // 尾部的near个
                        positions = (s_slc - near + (topKIdx - (topk - front - near)) - 1) * l_prime;
                    } else {
                        // 中间的topk-front-near个
                        SymbolicScalar topk_index;
                        if (debug) {
                            TileShape::Current().SetVecTile(1, 1, NUM16);
                            topk_index = GetTensorData(topK_indcies, {batchIdx, slcIdx, topKIdx - front});
                        } else {
                            topk_index = GetTensorData(topK_indcies, {batchIdx, slcIdx, topKIdx - front});
                        }
                        positions = topk_index * prime_value;
                    }
                    slcSeqLen = slcSeqLen + prime_value;
                    SymbolicScalar blockIdxInBatch;
                    SymbolicScalar tail;
                    if (blockSize != 0) {
                        blockIdxInBatch = positions / blockSize;
                        tail = positions % blockSize;
                    }
                    SymbolicScalar slcBlockIdx = GetTensorData(blockTable, {batchIdx, blockIdxInBatch});
                    TileShape::Current().SetVecTile(v0Tile[0], v0Tile[1]);
                    auto kv_slcBlock = View(kvNopeCache, {l_prime, kv_lora_rank}, {slcBlockIdx * blockSize + tail, nkvIdx * kv_lora_rank});
                    auto kRope_slcBlock = View(kRopeCache, {l_prime, rope_dim}, {slcBlockIdx * blockSize + tail, nkvIdx * rope_dim});
                    TileShape::Current().SetVecTile(v0Tile[0], v0Tile[1]);
                    auto kv_slcBlock_fp32 = Cast(kv_slcBlock, DataType::DT_FP32);
                    auto kRope_slcBlock_fp32 = Cast(kRope_slcBlock, DataType::DT_FP32);
                    TileShape::Current().SetVecTile(v0Tile[0], v0Tile[1]);
                    auto kv_slcBlock_tiled = MulS(kv_slcBlock_fp32, Element(kv_slcBlock_fp32->Datatype(), float(1)));
                    auto kRope_slcBlock_tiled = MulS(kRope_slcBlock_fp32, Element(kRope_slcBlock_fp32->Datatype(), float(1)));
                    TileShape::Current().SetVecTile(v0Tile[0], v0Tile[1]);
                    auto kv_slcBlock_fp16 = Cast(kv_slcBlock_tiled, k_slcOut->Datatype());
                    auto kRope_slcBlock_fp16 = Cast(kRope_slcBlock_tiled, v_slcOut->Datatype());
                    TileShape::Current().SetVecTile(v0Tile[0], v0Tile[1]);
                    SymbolicScalar output_axis1_value =
                    batchIdx * s * n2 * topk * l_prime + slcIdx * n2 * topk * l_prime + nkvIdx * topk * l_prime + topKIdx * l_prime;
                    Assemble(kv_slcBlock_fp16, {output_axis1_value, 0}, k_slcOut);
                    Assemble(kRope_slcBlock_fp16, {output_axis1_value, kv_lora_rank}, k_slcOut);
                    Assemble(kv_slcBlock_fp16, {output_axis1_value, 0}, v_slcOut);
                }
                SetTensorData(slcSeqLen, {batchIdx, slcIdx}, kvSlcActSeqs);
            }
        }
    }
}

}

int main() { // b_n_s_s2_h_q_lora_rank
    config::SetHostConfig(KEY_ONLY_CODEGEN, true);
    KvSlcTileShapeConfig tileConfig;
    const int nTile = 32;
    tileConfig.v0TileShape = {nTile, 32};

    std::vector<int> input_param = {1, 1, 1, 64, 64, 1, 2, 4, 16, 32};
    int b = input_param[0];
    int s = input_param[1];
    int n2 = input_param[2];
    int kv_lora_rank = input_param[3];
    int rope_dim = input_param[4];
    int front = input_param[5];
    int near = input_param[6];
    int topK = input_param[7];
    int l_prime = input_param[8];
    int blockSize = input_param[9];

    int blockNum = 0;
    std::vector<int> seq(NUM1024, 1);
    for (auto s_item : seq) {
        blockNum += CeilDiv(s_item, blockSize);
    }
    int maxSeqAllBatch = *(std::max_element(seq.begin(), seq.end()));
    int maxBlockNumPerBatch = CeilDiv(maxSeqAllBatch, blockSize);

    Tensor topk_tensor(DT_INT32, {b, s, topK - front - near}, "topk_tensor");
    Tensor topk_tensor_shape(DT_INT32, {b, s}, "topk_tensor_shape");
    Tensor kvNopeCache(DT_FP16, {int(blockNum * blockSize), n2 * kv_lora_rank}, "kNopeCache");
    Tensor kRopeCache(DT_FP16, {int(blockNum * blockSize), n2 * rope_dim}, "vNopeCache");
    Tensor kvActSeqs(DT_INT32, {b}, "kvActSeqs");
    Tensor blockTable(DT_INT32, {b, maxBlockNumPerBatch}, "blockTable");
    Tensor k_slcOut(DT_FP16, {b * s * n2 * topK * l_prime, rope_dim + kv_lora_rank}, "k_slcOut");
    Tensor v_slcOut(DT_FP16, {b * s * n2 * topK * l_prime, kv_lora_rank}, "v_slcOut");
    Tensor kvSlcActSeqs(DT_INT32, {b, s}, "kvSlcActSeqs");
    FUNCTION("main_slc",
        {topk_tensor, topk_tensor_shape, kvNopeCache, kRopeCache, kvActSeqs, blockTable}, {k_slcOut, v_slcOut, kvSlcActSeqs}) {
        KvSlcCompute(topk_tensor, topk_tensor_shape, kvNopeCache, kRopeCache, kvActSeqs, front, near, topK, l_prime, n2, blockTable, blockSize, k_slcOut, v_slcOut, kvSlcActSeqs, tileConfig);
    }
}

File: python/src/tensor_op/example/gen_kv_slc.py
#!/usr/bin/env python3
# coding: utf-8
# Copyright (c) 2025 Huawei Technologies Co., Ltd.
# This file is a part of the CANN Open Software.
# Licensed under CANN Open Software License Agreement Version 1.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# ======================================================================================================================
"""
"""
from dataclasses import dataclass
from typing import List
import logging
import math
import pto

TILE_VEC_DIMS = 2

NUM_16 = 16


@dataclass
class KvSlcTileShapeConfig:
    v0_tile_shape = [0] * TILE_VEC_DIMS


def kv_slc_compute(**kwargs):
    top_k_indcies: pto.tensor = kwargs.get("top_k_indcies")
    top_k_tensor_shape: pto.tensor = kwargs.get("top_k_tensor_shape")
    kv_nope_cache: pto.tensor = kwargs.get("kv_nope_cache")
    k_rope_cache: pto.tensor = kwargs.get("k_rope_cache")
    kv_act_seqs: pto.tensor = kwargs.get("kv_act_seqs")
    front: int = kwargs.get("front")
    near: int = kwargs.get("near")
    topk: int = kwargs.get("topk")
    l_prime: int = kwargs.get("l_prime")
    n2: int = kwargs.get("n2")
    block_table: pto.tensor = kwargs.get("block_table")
    block_size: int = kwargs.get("block_size")
    k_slc_out: pto.tensor = kwargs.get("k_slc_out")
    v_slc_out: pto.tensor = kwargs.get("v_slc_out")
    kv_slc_act_seqs: pto.tensor = kwargs.get("kv_slc_act_seqs")
    tile_config: KvSlcTileShapeConfig = kwargs.get("tile_config")
    debug: bool = kwargs.get("debug")

    v0_tile = tile_config.v0_tile_shape
    b = top_k_indcies.shape[0]
    s = top_k_indcies.shape[1]
    kv_lora_rank = kv_nope_cache.shape[1] // n2
    rope_dim = k_rope_cache.shape[1] // n2

    with pto.loop_function("LOOP_L0_batchIdx", "batch_idx", pto.loop_range(0, b, 1)) as batch_idx_loop:
        for batch_idx in batch_idx_loop:
            def inside_batch_idx_loop(batch_idx):
                cur_act_seq = pto.get_tensor_data(kv_act_seqs, [batch_idx])
                with pto.loop_function("LOOP_L1_slcIdx", "slc_idx", pto.loop_range(0, s, 1)) as slc_idx_loop:
                    for slc_idx in slc_idx_loop:
                        def inside_slc_idx_loop(slc_idx):
                            with pto.loop_function("LOOP_L2_kvSlcIdx", "nkv_idx",
                            pto.loop_range(0, n2, 1)) as nkv_idx_loop:
                                for nkv_idx in nkv_idx_loop:
                                    def inside_nkv_idx_loop(nkv_idx):
                                        pto.set_vec_tile_shapes(v0_tile[0], v0_tile[1])
                                        s_slc = pto.get_tensor_data(top_k_tensor_shape, [batch_idx, slc_idx])
                                        positions = 0
                                        prime_value = l_prime
                                        slc_seq_len = 0
                                        for top_k_idx in range(topk):
                                            if top_k_idx < front:
                                                positions = top_k_idx * l_prime
                                            else:
                                                topk_index = None
                                                if debug:
                                                    pto.set_vec_tile_shapes(1, 1, NUM_16)
                                                    topk_index = pto.get_tensor_data(top_k_indcies, [batch_idx, slc_idx,
                                                    pto.symbolic_scalar(top_k_idx - front)])
                                                else:
                                                    topk_index = pto.get_tensor_data(top_k_indcies, [batch_idx, slc_idx, 
                                                    pto.symbolic_scalar(top_k_idx - front)])
                                                positions = topk_index * prime_value
                                            slc_seq_len = slc_seq_len + prime_value
                                            block_idx_in_batch = positions // pto.symbolic_scalar(block_size)
                                            tail = positions % block_size
                                            slc_block_idx = pto.get_tensor_data(block_table, 
                                            [batch_idx, block_idx_in_batch])
                                            pto.set_vec_tile_shapes(v0_tile[0], v0_tile[1])
                                            kv_slc_block = pto.view(kv_nope_cache, [l_prime, kv_lora_rank],
                                            [slc_block_idx * block_size + tail, nkv_idx * kv_lora_rank])
                                            k_rope_slc_block = pto.view(k_rope_cache, [l_prime, rope_dim],
                                            [slc_block_idx * block_size + tail, nkv_idx * rope_dim])
                                            pto.set_vec_tile_shapes(v0_tile[0], v0_tile[1])
                                            kv_slc_block_fp32 = pto.cast(kv_slc_block, pto.DT_FP32)
                                            k_rope_slc_block_fp32 = pto.cast(k_rope_slc_block, pto.DT_FP32)
                                            pto.set_vec_tile_shapes(v0_tile[0], v0_tile[1])
                                            kv_slc_block_tiled = pto.mul_s(kv_slc_block_fp32,
                                            pto.element(kv_slc_block_fp32.get_dtype(), float(1)))
                                            k_rope_slc_block_tiled = pto.mul_s(k_rope_slc_block_fp32,
                                            pto.element(k_rope_slc_block_fp32.get_dtype(), float(1)))
                                            pto.set_vec_tile_shapes(v0_tile[0], v0_tile[1])
                                            kv_slc_block_fp16 = pto.cast(kv_slc_block_tiled, k_slc_out.get_dtype())
                                            k_rope_slc_block_fp16 = pto.cast(k_rope_slc_block_tiled,
                                            v_slc_out.get_dtype())
                                            pto.set_vec_tile_shapes(v0_tile[0], v0_tile[1])
                                            output_axis1_value = batch_idx * s * n2 * topk * l_prime
                                            + slc_idx * n2 * topk * l_prime
                                            + nkv_idx * topk * l_prime + top_k_idx * l_prime
                                            pto.assemble(kv_slc_block_fp16, [output_axis1_value, 0], k_slc_out)
                                            pto.assemble(k_rope_slc_block_fp16,
                                            [output_axis1_value, kv_lora_rank], k_slc_out)
                                            pto.assemble(kv_slc_block_fp16, [output_axis1_value, 0], v_slc_out)
                                        pto.set_tensor_data(pto.symbolic_scalar(slc_seq_len),
                                        [batch_idx, slc_idx], kv_slc_act_seqs)
                                    inside_nkv_idx_loop(slc_idx)
                        inside_slc_idx_loop(slc_idx)
            inside_batch_idx_loop(batch_idx)


if __name__ == '__main__':
    tile_config = KvSlcTileShapeConfig()
    n_tile = 32
    tile_config.v0_tile_shape = [n_tile, 32]

    input_param: List[int] = [1, 1, 1, 64, 64, 1, 2, 4, 16, 32]
    b = input_param[0]
    s = input_param[1]
    n2 = input_param[2]
    kv_lora_rank = input_param[3]
    rope_dim = input_param[4]
    front = input_param[5]
    near = input_param[6]
    topk = input_param[7]
    l_prime = input_param[8]
    block_size = input_param[9]

    block_num = 0
    seq = [1] * 1024
    for s_item in seq:
        block_num += (s_item + block_size - 1) // block_size
    max_seq_all_batch = max(seq)
    max_block_num_per_batch = math.ceil(max_seq_all_batch / block_size)

    top_k_indcies = pto.tensor([b, s, topk - front - near], pto.DT_INT32, "top_k_indcies")
    top_k_tensor_shape = pto.tensor([b, s], pto.DT_INT32, "top_k_tensor_shape")
    kv_nope_cache = pto.tensor([int(block_num * block_size), n2 * kv_lora_rank], pto.DT_FP16, "kv_nope_cache")
    k_rope_cache = pto.tensor([int(block_num * block_size), n2 * rope_dim], pto.DT_FP16, "k_rope_cache")
    kv_act_seqs = pto.tensor([b], pto.DT_INT32, "kv_act_seqs")
    block_table = pto.tensor([b, max_block_num_per_batch], pto.DT_INT32, "block_table")
    k_slc_out = pto.tensor([b * s * n2 * topk * l_prime, rope_dim + kv_lora_rank], pto.DT_FP16, "k_slc_out")
    v_slc_out = pto.tensor([b * s * n2 * topk * l_prime, kv_lora_rank], pto.DT_FP16, "v_slc_out")
    kv_slc_act_seqs = pto.tensor([b, s], pto.DT_INT32, "kv_slc_act_seqs")
    input_tensors = [
        top_k_indcies, top_k_tensor_shape, kv_nope_cache,
        k_rope_cache, kv_act_seqs, block_table
        ]
    output_tensors = [k_slc_out, v_slc_out, kv_slc_act_seqs]
    with pto.function("main_slc", input_tensors, output_tensors):
        kv_slc_compute(
            top_k_indcies=top_k_indcies,
            top_k_tensor_shape=top_k_tensor_shape,
            kv_nope_cache=kv_nope_cache,
            k_rope_cache=k_rope_cache,
            kv_act_seqs=kv_act_seqs,
            front=front,
            near=near,
            topk=topk,
            l_prime=l_prime,
            n2=n2,
            block_table=block_table,
            block_size=block_size,
            k_slc_out=k_slc_out,
            v_slc_out=v_slc_out,
            kv_slc_act_seqs=kv_slc_act_seqs,
            tile_config=tile_config
        )
